File: rust/wasplib/src/client/keys.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use super::host::get_key_id_from_string;

// @formatter:off
pub const KEY_AGENT       : i32 =                 -1;
pub const KEY_AMOUNT      : i32 = KEY_AGENT       -1;
pub const KEY_BALANCES    : i32 = KEY_AMOUNT      -1;
pub const KEY_BASE58      : i32 = KEY_BALANCES    -1;
pub const KEY_CALLER      : i32 = KEY_BASE58      -1;
pub const KEY_CALLS       : i32 = KEY_CALLER      -1;
pub const KEY_CHAIN       : i32 = KEY_CALLS       -1;
pub const KEY_CHAIN_OWNER : i32 = KEY_CHAIN       -1;
pub const KEY_COLOR       : i32 = KEY_CHAIN_OWNER -1;
pub const KEY_CONTRACT    : i32 = KEY_COLOR       -1;
pub const KEY_CREATOR     : i32 = KEY_CONTRACT    -1;
pub const KEY_DATA        : i32 = KEY_CREATOR     -1;
pub const KEY_DELAY       : i32 = KEY_DATA        -1;
pub const KEY_DESCRIPTION : i32 = KEY_DELAY       -1;
pub const KEY_ERROR       : i32 = KEY_DESCRIPTION -1;
pub const KEY_EVENT       : i32 = KEY_ERROR       -1;
pub const KEY_EXPORTS     : i32 = KEY_EVENT       -1;
pub const KEY_FUNCTION    : i32 = KEY_EXPORTS     -1;
pub const KEY_HASH        : i32 = KEY_FUNCTION    -1;
pub const KEY_ID          : i32 = KEY_HASH        -1;
pub const KEY_INCOMING    : i32 = KEY_ID          -1;
pub const KEY_IOTA        : i32 = KEY_INCOMING    -1;
pub const KEY_LENGTH      : i32 = KEY_IOTA        -1;
pub const KEY_LOG         : i32 = KEY_LENGTH      -1;
pub const KEY_LOGS        : i32 = KEY_LOG         -1;
pub const KEY_NAME        : i32 = KEY_LOGS        -1;
pub const KEY_PANIC       : i32 = KEY_NAME        -1;
pub const KEY_PARAMS      : i32 = KEY_PANIC       -1;
pub const KEY_POSTS       : i32 = KEY_PARAMS      -1;
pub const KEY_RANDOM      : i32 = KEY_POSTS       -1;
pub const KEY_RESULTS     : i32 = KEY_RANDOM      -1;
pub const KEY_STATE       : i32 = KEY_RESULTS     -1;
pub const KEY_TIMESTAMP   : i32 = KEY_STATE       -1;
pub const KEY_TRACE       : i32 = KEY_TIMESTAMP   -1;
pub const KEY_TRANSFERS   : i32 = KEY_TRACE       -1;
pub const KEY_UTILITY     : i32 = KEY_TRANSFERS   -1;
pub const KEY_VIEWS       : i32 = KEY_UTILITY     -1;
pub const KEY_WARNING     : i32 = KEY_VIEWS       -1;
pub const KEY_ZZZZZZZ     : i32 = KEY_WARNING     -1;
// @formatter:on

pub trait MapKey {
    fn get_id(&self) -> i32;
}

impl MapKey for str {
    fn get_id(&self) -> i32 {
        get_key_id_from_string(self)
    }
}

impl MapKey for i32 {
    fn get_id(&self) -> i32 {
        *self
    }
}

File: packages/vm/wasmlib/keys.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

type MapKey interface {
	KeyID() Key32
}

type Key string

func (key Key) KeyID() Key32 {
	return GetKeyIDFromString(string(key))
}

type Key32 int32

func (key Key32) KeyID() Key32 {
	return key
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const (
	KeyAccountID       = Key32(-1)
	KeyAddress         = Key32(-2)
	KeyBalances        = Key32(-3)
	KeyBase58Decode    = Key32(-4)
	KeyBase58Encode    = Key32(-5)
	KeyBlsAddress      = Key32(-6)
	KeyBlsAggregate    = Key32(-7)
	KeyBlsValid        = Key32(-8)
	KeyCall            = Key32(-9)
	KeyCaller          = Key32(-10)
	KeyChainID         = Key32(-11)
	KeyChainOwnerID    = Key32(-12)
	KeyColor           = Key32(-13)
	KeyContract        = Key32(-14)
	KeyContractCreator = Key32(-15)
	KeyDeploy          = Key32(-16)
	KeyEd25519Address  = Key32(-17)
	KeyEd25519Valid    = Key32(-18)
	KeyEvent           = Key32(-19)
	KeyExports         = Key32(-20)
	KeyHashBlake2b     = Key32(-21)
	KeyHashSha3        = Key32(-22)
	KeyHname           = Key32(-23)
	KeyIncoming        = Key32(-24)
	KeyLength          = Key32(-25)
	KeyLog             = Key32(-26)
	KeyMaps            = Key32(-27)
	KeyMinted          = Key32(-28)
	KeyName            = Key32(-29)
	KeyPanic           = Key32(-30)
	KeyParams          = Key32(-31)
	KeyPost            = Key32(-32)
	KeyRandom          = Key32(-33)
	KeyRequestID       = Key32(-34)
	KeyResults         = Key32(-35)
	KeyReturn          = Key32(-36)
	KeyState           = Key32(-37)
	KeyTimestamp       = Key32(-38)
	KeyTrace           = Key32(-39)
	KeyTransfers       = Key32(-40)
	KeyUtility         = Key32(-41)
	KeyValid           = Key32(-42)
	KeyZzzzzzz         = Key32(-43)
)
